trieste/models/gpflow/architectures.py

```diff
@@ -0,0 +1,159 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" This file contains builders for some GPflow models supported in Trieste. """
+
+from __future__ import annotations
+
+from typing import Optional
+
+import gpflow
+import tensorflow as tf
+import tensorflow_probability as tfp
+from gpflow.models import GPR, SGPR, SVGP, VGP
+
+from ...data import Dataset
+from ...space import Box
+
+
+def build_gpr(data: Dataset, trainable_likelihood: bool = False, kernel_priors: bool = True) -> GPR:
+    """
+    Constructs a :class:`~gpflow.models.GPR` model with sensible initial parameters and
+    (optionally) priors. We use :class:`~gpflow.kernels.Matern52` kernel in the model as we found
+    it to be effective in most settings. We assume inputs are normalised to the unit hypercube for
+    setting sensible initial parameters for the kernel.
+
+    We set priors for kernel hyperparameters by default in order to stabilize model fitting. We
+    found the priors below to be highly effective for objective functions defined over the unit
+    hypercube. For objective functions with different scaling, other priors will likely be more
+    appropriate. Using priors allows for using maximum a posteriori estimate of these kernel
+    parameters during model fitting.
+
+    :param data: Dataset from the initial design, used for estimating the variance of observations.
+    :param trainable_likelihood: If set to `True` likelihood paramater is set to
+        trainable. By default set to `False`.
+    :param priors: If set to `True` (default) priors are set for kernel parameters.
+    :return: A :class:`~gpflow.models.GPR` model.
+    """
+    empirical_variance = tf.math.reduce_variance(data.observations)
+    dim = data.query_points.shape[-1]
+
+    kernel = gpflow.kernels.Matern52(variance=empirical_variance, lengthscales=[0.2] * dim)
+
+    if kernel_priors:
+        prior_scale = tf.cast(1.0, dtype=gpflow.default_float())
+        kernel.variance.prior = tfp.distributions.LogNormal(
+            tf.math.log(empirical_variance), prior_scale
+        )
+        kernel.lengthscales.prior = tfp.distributions.LogNormal(
+            tf.math.log(kernel.lengthscales), prior_scale
+        )
+
+    model = gpflow.models.GPR(data.astuple(), kernel, noise_variance=1e-5)
+
+    if not trainable_likelihood:
+        gpflow.set_trainable(model.likelihood, False)
+
+    return model
+
+
+def build_sgpr(
+    data: Dataset,
+    search_space: Optional[Box] = None,
+    trainable_likelihood: bool = False,
+    trainable_inducing_points: bool = False,
+) -> SGPR:
+    empirical_variance = tf.math.reduce_variance(data.observations)
+    dim = data.query_points.shape[-1]
+    num_data = len(data.observations)
+
+    kernel = gpflow.kernels.Matern52(variance=empirical_variance, lengthscales=[0.2] * dim)
+
+    if search_space:
+        inducing_points = search_space.sample_sobol(num_data)
```

Comment:
I think that this is a good default for initialisation. 

With @henrymoss we found that the best strategy was to use k-means, but that only truly matters when we update the model. So not sure k-means belongs here (question remains: should we have a good update default for inducings?).

Reply:
would you make it equal to number of initial data points or relax it to some number?

Reply:
The efficient way of dealing with sparse models in trieste is to use a fixed number of inducing points (if you change this number your trigger a lot of retracing).

We should come up with a heuristic formula, e.g. growing the number with the dimension. In 2D we could use something like 50 points, and we should never use more than 500 points. How about 

`num_inducings = min(500, 25 *dim)? `

and we also make this an optional input so it's easy to use something else.

Reply:
sounds good... you would keep only sobol inducing points, not having an option of setting them to the initial data?

Reply:
yes I think this is fine. Possibly we'll update this in the future if we have good evidence of a better strategy?